[sending] Add support for 'sending'

gottesmm opened the pull request:
This PR adds support to the compiler for 'sending'. It does not completely remove 'transferring' (since it may already be used). Instead what it does is:

1. Internally except for in the parser and the clang importer, we only represent
'sending'. This means that it will be easy to remove 'transferring' once enough
time has passed.

2. I included a warning that suggested to the user to change 'transferring' ->
'sending'.

3. I duplicated the parsing diagnostics for 'sending' so both will still get
different sets of diagnostics for parsing issues... but anywhere below parsing,
I have just changed 'transferring' to 'sending' since transferring isn't
represented at those lower levels.

4. Since SendingArgsAndResults is always enabled when TransferringArgsAndResults
is enabled (NOTE not vis-a-versa), we know that we can always parse sending. So
we import "transferring" as "sending". This means that even if one marks a
function with "transferring", the compiler will guard it behind a
SendingArgsAndResults -D flag and in the imported header print out sending.

rdar://128216574

gottesmm committed:
[sending] Change the internals of sending to be based around 'sending' instead of 'transferring'.

We still only parse transferring... but this sets us up for adding the new
'sending' syntax by first validating that this internal change does not mess up
the current transferring impl since we want both to keep working for now.

rdar://128216574

gottesmm committed:
[sending] Change transferring diagnostics to say sending instead of transferring.

Just trying to slice off a larger change where I change these tests to actually
use 'sending'. This is nice to do now since it is algebraic to do.

rdar://128216574

gottesmm committed:
[sending] Begin parsing 'sending' while still accepting 'transferring'.

A few things:

1. Internally except for in the parser and the clang importer, we only represent
'sending'. This means that it will be easy to remove 'transferring' once enough
time has passed.

2. I included a warning that suggested to the user to change 'transferring' ->
'sending'.

3. I duplicated the parsing diagnostics for 'sending' so both will still get
different sets of diagnostics for parsing issues... but anywhere below parsing,
I have just changed 'transferring' to 'sending' since transferring isn't
represented at those lower levels.

4. Since SendingArgsAndResults is always enabled when TransferringArgsAndResults
is enabled (NOTE not vis-a-versa), we know that we can always parse sending. So
we import "transferring" as "sending". This means that even if one marks a
function with "transferring", the compiler will guard it behind a
SendingArgsAndResults -D flag and in the imported header print out sending.

rdar://128216574

gottesmm committed:
[sending] Rename some tests from transferring -> sending.

I am doing this after the main transformation to ease commit by commit
understanding of the change.

rdar://128216574

gottesmm commented:
@swift-ci smoke test

hborla commented on include/swift/Basic/Features.def:
```diff
@@ -353,7 +353,14 @@ EXPERIMENTAL_FEATURE(FixedArrays, true)
+EXPERIMENTAL_FEATURE(TransferringArgsAndResults, true)
+
+// Allow for the 'sending' keyword to be applied to arguments and results.
+SUPPRESSIBLE_EXPERIMENTAL_FEATURE(SendingArgsAndResults, true)
```
These should not be experimental features

gottesmm committed:
[sending] Enable sending whenever region isolation is enabled and make SE-0430 an upcoming feature.

I also added support for expressing suppressable upcoming features.

rdar://128216574

ktoso reviewed (approved):
LGTM, when do you think we'll remove `transfering` - will we be able to drop it in the final 6.0 or may have to tag it along longer? 🤔 

Cool thx, when this lands I'll follow up with some changes in distributed

ahoppen commented:
I think this also needs a corresponding parser change in swift-syntax.

gottesmm commented:
apple/swift-syntax#2658

@swift-ci smoke test

gottesmm commented:
https://github.com/apple/swift-syntax/pull/2658

@swift-ci smoke test
